### PR TITLE
Default Components folder will be created at init time (#340)

### PR DESCRIPTION
**Pull request by vinayada1**
# Description

- Dapr init will create a components folder under homedir/.dapr
- Dapr run will use this unless specified by --components-path
- Dapr uninstall will delete the default components folder

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #340 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ *] Code compiles correctly
* [ *] Created/updated tests
* [ *] Extended the documentation

### CODE REVIEW

File: pkg/standalone/common.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	"runtime"
+)
+
+// getDefaultComponentsFolder returns the hidden .components folder created under install directory at init time
+func getDefaultComponentsFolder() string {
+	const daprDirName = ".dapr"
+	const componentsDirName = "components"
+	daprDirPath := os.Getenv("HOME")
+	if runtime.GOOS == daprWindowsOS {
+		daprDirPath = path_filepath.Join(os.Getenv("HOMEDRIVE"), os.Getenv("HOMEPATH"))
```
Comment (LMWF) [inline pkg/standalone/common.go:15]: The user profile env var should work here.

File: pkg/standalone/run.go [hunk 2]
```diff
@@ -369,3 +343,13 @@ func Run(config *RunConfig) (*RunOutput, error) {
...
 	}, nil
 }
+
+func getComponentsPath(config *RunConfig) (string, error) {
+	if config.ComponentsPath == "" {
+		componentsPath := getDefaultComponentsFolder()
+		return componentsPath, nil
+	}
+
+	_, err := os.Stat(config.ComponentsPath)
```
Comment (LMWF) [inline pkg/standalone/run.go:353]: This should check if the path exists as well.  The usage could clarify if the cli expects the specified path to exist ahead of time or if the cli will create it.

File: pkg/standalone/uninstall.go [hunk 3]
```diff
@@ -62,6 +70,11 @@ func Uninstall(uninstallAll bool, dockerNetwork string) error {
 		fmt.Println("WARNING: could not delete run data file")
 	}
 
+	err = removeDefaultComponentsFolder()
+	if err != nil {
+		fmt.Println("WARNING: could not delete default components folder")
```
Comment (mukundansundar) [inline pkg/standalone/uninstall.go:75]: probably print location of the default components folder here?

File: utils/utils.go [hunk 4]
```diff
@@ -19,6 +19,9 @@ import (
 	"github.com/olekukonko/tablewriter"
 )
 
+// ComponentsDirName is the default hidden components folder name created at init time
+const ComponentsDirName = ".components"
```
Comment (mukundansundar) [inline utils/utils.go:23]: not used

File: pkg/standalone/common.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package standalone
+
+import (
+	"os"
+	path_filepath "path/filepath"
+	"runtime"
+)
+
+// getDefaultComponentsFolder returns the hidden .components folder created under install directory at init time
```
Comment (mukundansundar) [inline pkg/standalone/common.go:9]: nitpick: change the comment: `returns the path to default components folder under home directory at init`

File: README.md [hunk 6]
```diff
@@ -60,7 +60,7 @@ $ dapr init
 ✅  Success! Dapr is up and running
 ```
 
-> Note: To see that Dapr has been installed successful, from a command prompt run the `docker ps` command and check that the `daprio/dapr:latest` and `redis` container images are both running.
+> Note: To see that Dapr has been installed successful, from a command prompt run the `docker ps` command and check that the `daprio/dapr:latest` and `redis` container images are both running. Also, this step creates a components folder under your home directory/.dapr/components which is later used at runtime
```
Comment (pruthvidhodda) [inline README.md:63]: nit: successful --> successfully

File: pkg/standalone/standalone.go [hunk 7]
```diff
@@ -389,6 +389,21 @@ func installDaprBinary(wg *sync.WaitGroup, errorChan chan<- error, dir, version
 	errorChan <- nil
 }
 
+func createComponentsDir(wg *sync.WaitGroup, errorChan chan<- error, dir, version string, dockerNetwork string, installLocation string) {
+	defer wg.Done()
+
+	// Make default components directory under install path
```
Comment (pruthvidhodda) [inline pkg/standalone/standalone.go:395]: This is not the install path

File: pkg/standalone/standalone_test.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
+// +build large_timeout
```
[the remark this answers is not in the document]
Comment (vinayada1) [reply pkg/standalone/standalone_test.go]: Removed the tests

File: pkg/standalone/common.go (hunk 1 above)
Comment (vinayada1) [reply pkg/standalone/common.go:15]: fixed

File: pkg/standalone/uninstall.go (hunk 3 above)
Comment (vinayada1) [reply pkg/standalone/uninstall.go:75]: done

File: pkg/standalone/common.go (hunk 5 above)
Comment (vinayada1) [reply pkg/standalone/common.go:9]: fixed

File: README.md (hunk 6 above)
Comment (vinayada1) [reply README.md:63]: fixed

File: pkg/standalone/standalone.go (hunk 7 above)
Comment (vinayada1) [reply pkg/standalone/standalone.go:395]: fixed

File: pkg/standalone/run.go (hunk 2 above)
Comment (vinayada1) [reply pkg/standalone/run.go:353]: Not sure if I understood you. Updated both code paths to test for dir to exist and added more doc in the README to clarify. Also, the idea behind the default components folder is that the new users might not even know what the components folder is for and we create it for them to help them get started.

File: utils/utils.go (hunk 4 above)
Comment (vinayada1) [reply utils/utils.go:23]: fixed

File: pkg/standalone/run.go (hunk 2 above)
Comment (LMWF) [reply pkg/standalone/run.go:353]: > Also, the idea behind the default components folder is that the new users might not even know what the components folder is for and we create it for them to help them get started.

Yes I mean in the case where the user specified --components-path.

Comment (vinayada1) [reply pkg/standalone/run.go:353]: Added

File: pkg/standalone/run.go [hunk 9]
```diff
@@ -374,3 +348,13 @@ func Run(config *RunConfig) (*RunOutput, error) {
...
+	return config.ComponentsPath, err
```
Comment (mukundansundar) [inline pkg/standalone/run.go:359]: is the expected behavior that the `config.ComponentsPath` exists previously and not created on the fly as was the previous behavior?

File: README.md [hunk 10]
```diff
@@ -60,7 +60,7 @@ $ dapr init
 ✅  Success! Dapr is up and running
 ```
 
-> Note: To see that Dapr has been installed successful, from a command prompt run the `docker ps` command and check that the `daprio/dapr:latest` and `redis` container images are both running.
+> Note: To see that Dapr has been installed successfully, from a command prompt run the `docker ps` command and check that the `daprio/dapr:latest` and `redis` container images are both running. Also, this step creates a default components folder under your home directory/.dapr/components which is later used at runtime unless the --components-path option is provided
```
Comment (amanbha) [inline README.md:63]: Clarify the paths for windows and Linux /MacOs here

File: pkg/standalone/run.go (hunk 9 above)
Comment (vinayada1) [reply pkg/standalone/run.go:359]: The path was created during dapr run before. Now it will be created at dapr init time

File: README.md (hunk 10 above)
Comment (vinayada1) [reply README.md:63]: Added

File: README.md [hunk 11]
```diff
@@ -60,7 +60,7 @@ $ dapr init
 ✅  Success! Dapr is up and running
 ```
 
-> Note: To see that Dapr has been installed successful, from a command prompt run the `docker ps` command and check that the `daprio/dapr:latest` and `redis` container images are both running.
+> Note: To see that Dapr has been installed successfully, from a command prompt run the `docker ps` command and check that the `daprio/dapr:latest` and `redis` container images are both running. Also, this step creates a default components folder which is later used at runtime unless the --components-path option is provided. For Linux/MacOS, the default components folder path is ~/.dapr/components and for Windows it is %USERPROFILE%\.dapr\components.
```
Comment (amanbha) [inline README.md:63]: for linux it would be $HOME/.dapr/components

**Comment by yaron2** [comment]
We need to issue a docs update in the docs and samples repos that explain this new behavior.

**Comment by vinayada1** [comment]
> We need to issue a docs update in the docs and samples repos that explain this new behavior.

I have these PRs open:
https://github.com/dapr/samples/pull/228
https://github.com/dapr/docs/pull/621

**Pull request merged**